src/react/component-name.ts

```diff
@@ -0,0 +1,44 @@
+import {readFileSync} from 'fs'
+import * as R from 'ramda'
+import {RuleParamsBase} from './types'
+import {filterPaths, getUniquePaths} from './utils'
+
+/**
+ * For all created/modified files traverses up through all containing folders
+ * and requires the name of the exported component constant to match the name of the function in the memo.
+ *
+ * @param danger Danger instance
+ * @param fail Danger fail function
+ * @param excludePaths paths to exclude
+ * @param includePaths paths to include
+ */
+export const componentName = (params: RuleParamsBase): void => {
+  R.compose(
+    R.forEach((path: string) => {
+      const {fail} = params
+
+      let matches: RegExpMatchArray | null | undefined
+      try {
+        matches = readFileSync(`${path}/index.tsx`, {encoding: 'utf8', flag: 'r'}).match(
+          /(\w+) = memo\(function (\w+)\(/i,
+        )
+      } catch (error: any) {
+        // Any component's dir must have index.tsx within it. If index.tsx file was not found then it is not a component's dir
+        return
+      }
+
+      if (!matches) {
+        return
+      }
+
+      const [, _componentName, functionName] = matches
```

Comment:
@quesabe why is the `_componentName` underlined?

Reply:
It clashes with the function name (the exported rule). My attempts to invent something better led to long and verbose names, thus I ended up with. just an underscore.